services/gem/gem-downloads.service.js

```diff
@@ -0,0 +1,149 @@
+'use strict'
+
+const semver = require('semver')
+const Joi = require('joi')
+
+const { BaseJsonService } = require('../base')
+const { InvalidResponse } = require('../errors')
+const {
+  downloadCount: downloadCountColor,
+} = require('../../lib/color-formatters')
+const { metric } = require('../../lib/text-formatters')
+const { latest: latestVersion } = require('../../lib/version')
+const { positiveInteger } = require('../validators.js')
+
+const downloadsSchema = Joi.alternatives().try(
+  Joi.object({
+    downloads: positiveInteger,
+    version_downloads: positiveInteger,
+  }).required(),
+  Joi.array()
+    .items(
+      Joi.object({
+        prerelease: Joi.boolean().required(),
+        number: Joi.string().required(),
+        downloads_count: positiveInteger,
+      })
+    )
+    .min(1)
+    .required()
```

Comment:
I hit an edge case here. Depending on which downloads badge we are building (`dt`/`dtv` or `dv`) we request a different endpoint on `rubygems.org`. This means there is more than one possible response schema. We can cover it by using `Joi.alternatives().try()` like I've done here, but I wonder if it could be better to think in terms of a URL pattern/schema pair here instead of a one-to-one badge/schema relationship? Any thoughts? Am I just over-complicating things? Are there other situations where we might hit a more complicated version of this?

Reply:
Yea, the npm downloads badge had the same issue. There I handled it with a switch: https://github.com/badges/shields/blob/2ebb173a62bb6f291104920ed4ed8a9a38934217/services/npm/npm-downloads.service.js#L46

Reply:
Yes - that's a much better way of doing it